core-addon/IonCore.js

```diff
@@ -60,6 +66,9 @@ module.exports = class IonCore {
         // is expecting it.
         return this._enrollStudy(message.data.studyID).then(r => true);
       } break;
+      case "unenrollment": {
+        return this._unenroll().then(r => true);
```

Comment:
I'm a little confused by this idiom... is this saying that if `_unenroll()` doesn't throw anything then return `true` essentially? I think it's OK but seems like we really want `initialize()` to be an `async` function or something.

Since there's no `catch` then this means the exception is basically unhandled?

Reply:
> 
> 
> I'm a little confused by this idiom... is this saying that if `_unenroll()` doesn't throw anything then return `true` essentially?

I'm saying "only return true if `_unenroll` is successful". On problems, the exception bubbles up to the UX side.

>I think it's OK but seems like we really want `initialize()` to be an `async` function or something.

Why do we need `initialize()` to be `async`? Did you mean `_handleMessage`? This function is basically always returning a promise (either resolved or rejected). The result of such promise is returned to the sender. This is the [recommended way advertised on our MDN docs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage).

> Since there's no `catch` then this means the exception is basically unhandled?

Good catch. This is basically saying "it's a UX problem". But you nailed the underlying bigger problem: how are we handling errors or things that go wrong on the UI? The current approach makes the UI aware, but this means that the UI will break if no error is handled.

Should we file an issue and deal about this problem separately? It seems like a big topic on its own.

Reply:
Created #37!